index.bs

```diff
@@ -9,8 +9,184 @@ Editor: Kent Tamura, Google https://www.google.com/, tkent@google.com
+    shadowRoot.innerHTML =
+      &#96;&lt;style>
+       :host::before {
+         content: '[ ]';
+         white-space: pre;
+         font-family: monospace;
+       }
+       :host(:state(checked))::before { content: '[x]' }
+       &lt;/style>
+       &lt;slot>Label&lt;/slot>&#96;;
+  }
+
+  get checked() { return this._internals.states.contains('checked'); }
+
+  set checked(flag) {
+    this._internals.states.toggle('checked', !!flag);
+  }
+
+  _onClick(event) {
+    this.checked = !this.checked;
+  }
+}
+
+customElements.define('labeled-checkbox', LabeledCheckbox);
+&lt;/script>
+
+&lt;style>
+labeled-checkbox { border: dashed red; }
+labeled-checkbox:state(checked) { border: solid; }
+&lt;/style>
+
+&lt;labeled-checkbox>You need to check this&lt;/labeled-checkbox>
+
+<!-- Works even on ::part()s -->
+&lt;script>
+class QuestionBox extends HTMLElement {
+  constructor() {
+    super();
+    const shadowRoot = this.attachShadow({mode: 'closed'});
+    shadowRoot.innerHTML =
+      &#96;&lt;div>&lt;slot>Question&lt;/slot>&lt;/div>
+       &lt;labeled-checkbox part='checkbox'>Yes&lt;/labeled-checkbox>&#96;;
+  }
+}
+customElements.define('question-box', QuestionBox);
+&lt;/script>
+
+&lt;style>
+question-box::part(checkbox) { color: red; }
+question-box::part(checkbox):state(checked) { color: green; }
+&lt;/style>
+
+&lt;question-box>Continue?&lt;/question-box>
+&lt;/body>
+</pre>
+</div>
+
+
+Exposing custom element states {#exposing}
+============================
+
+Each <a>autonomous custom element</a> has <dfn>custom states</dfn>, a [=/set=]
+of strings. It's an empty set initially.
+
+<pre class=idl>
+partial interface ElementInternals {
+  [SameObject, PutForwards=value] readonly attribute DOMTokenList states;
+};
+</pre>
+
+The {{states}} IDL attribute returns a [=non-attribute DOMTokenList=] object
+associated with context object's
+<a href="http://html.spec.whatwg.org/C/#internals-target">target element</a>
+and its [=custom states=].
+
+<div class="example" id="ex-non-boolean-state">
+[=Custom states=] can expose boolean states represented by
+existence/non-exsitence of string tokens. If an author wants to expose a state
+which can have three values, it can be converted to three boolean states.
+For example, a state called <code>readyState</code> with <code>"loading"</code>,
+<code>"interactive"</code>, and <code>"complete"</code> values can be mapped to
+three boolean states, <code>"readyState_loading"</code>,
+<code>"readyState_interactive"</code>, and <code>"readyState_complete"</code>.
+
+<pre class="lang-js">
+// Change the readyState from anything to "complete".
+this.readyState = "complete";
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/remove}}</l>("readyState_loading", "readyState_interactive");
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/add}}</l>("readyState_complete");
+</pre>
+</div>
+
+## Non-attribute DOMTokenList ## {#sec-non-attribute-domtokenlist}
```

Comment:
This section is reasonable for now but I think eventually it should be done as a patch to the DOM Standard that adds internal branches to each part of the DOMTokenList spec. I think that will force a bit more rigor than we have here.

For example, I would do something like modifying the current sentence in DOM that says

> A DOMTokenList object also has an associated element and an attribute’s local name. 

to

> A DOMTokenList object also has an associated element and an optional attribute’s local name. If the DOMTokenList object has an attribute's local name, then it is known as an **attribute-associated `DOMTokenList`**; otherwise it is a **non-attribute `DOMTokenList`**.

Reply:
I totally agree that this section should be merged into the DOM specification.
Do you have any idea about when we should merge?

Reply:
Probably around the same time we merge the rest of the spec into Selectors/HTML, i.e. when it has multi-implementer interest and is ready to graduate from incubation. There is no rush.

What I meant was that you could write this section as if it were a patch to DOM, even if for now it lives in this document.

Reply:
Ah, now I understand the intention of your previous comment.  I updated this section like a patch.